crates/oxc_linter/src/rules/eslint/no_extra_boolean_cast.rs

```diff
@@ -0,0 +1,879 @@
+struct NoExtraDoubleNegationCastDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-extra-boolean-cast): Redundant Boolean call")]
+#[diagnostic(severity(warning), help("Remove the Boolean call as it will be coerced to boolean"))]
+struct NoExtraBooleanCastDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoExtraBooleanCast {
+    pub enforce_for_logical_operands: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// This rule disallows unnecessary boolean casts.
+    ///
+    /// ### Why is this bad?
+    /// In contexts such as an if statement’s test where the result of the expression will already be coerced to a Boolean,
+    /// casting to a Boolean via double negation (!!) or a Boolean call is unnecessary.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var foo = !!!bar;
+    /// var foo = Boolean(!!bar);
+    ///
+    /// if (!!foo) {}
+    /// if (Boolean(foo)) {}
+    ///
+    /// // with "enforceForLogicalOperands" option enabled
+    /// if (!!foo || bar) {}
+    /// ```
+    NoExtraBooleanCast,
+    correctness
+);
+
+impl Rule for NoExtraBooleanCast {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        Self {
+            enforce_for_logical_operands: value
+                .get(0)
+                .and_then(|x| x.get("enforceForLogicalOperands"))
+                .and_then(serde_json::Value::as_bool)
+                .unwrap_or(false),
+        }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::CallExpression(expr) = node.kind() {
+            if let Some(ident) = expr.callee.without_parenthesized().get_identifier_reference() {
+                if ident.name != "Boolean" {
+                    return;
+                }
+            }
+            if is_flagged_ctx(node, ctx, self.enforce_for_logical_operands) {
+                ctx.diagnostic(NoExtraBooleanCastDiagnostic(expr.span));
+            }
+        }
+
+        let parent = get_real_parent(node, ctx);
+        if parent.is_none() {
+            return;
+        }
+        let parent = parent.unwrap();
+
+        if let (AstKind::UnaryExpression(expr), AstKind::UnaryExpression(parent_expr)) =
+            (node.kind(), parent.kind())
+        {
+            match (expr.operator, parent_expr.operator) {
+                (UnaryOperator::LogicalNot, UnaryOperator::LogicalNot)
+                    if is_flagged_ctx(parent, ctx, self.enforce_for_logical_operands) =>
+                {
+                    ctx.diagnostic(NoExtraDoubleNegationCastDiagnostic(expr.span));
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+// Checks whether the node is a context that should report an error
+// Acts recursively if it is in a logical context
+fn is_flagged_ctx(node: &AstNode, ctx: &LintContext, enforce_for_logical_operands: bool) -> bool {
```

Comment:
This is a worst-case triple* iteration.
1. `is_bool_context` invokes `get_real_parent`, which iterates up the ast
2. `get_real_context` is called (if above branch evaluates to `false`
3. Assuming that branch evaluates to true, the process repeats on the parent

Consider finding a shorter, version of this algorithm that requires viewer AST traversals